rfcs/2021-07-23-8447-extract-helm-charts.md

```diff
@@ -0,0 +1,119 @@
+It's not a requirement to automate these steps and scripts/make targets can be used instead if we decide to reduce scope and automate
+these tasks at a later time.
+
+The `k8s-test-framework` already contains an `external_chart` method, used for deploying DataDog's agent chart during integration tests.
+This allows us to update tests that expect Vector's charts to be located in the same repository to instead reference our hosted charts
+at https://packages.timber.io/helm.
+
+Issues directly related to Vector's Helm charts (or the resources generated from them) will be maintained in the `timberio/helm-charts`
+repository. Existing issues will be migrated as needed.
+
+## Rationale
+
+This change aligns us with standard tooling and practices in the Helm community, not to mention DataDog's existing workflows. The
+immediate benefit to contributors would be drastically shrinking the CI feedback loop by no longer requiring the compilation and image
+building of Vector to test changes to the chart templates and configuration. This in turn allows us to start reducing the responsibilities
+of the existing integration tests, which reduces the feedback time for changes to Vector that actually require integration testing on Kubernetes.
+
+## Prior Art
+
+### DataDog Agent
+
+Helm charts are stored separately in [DataDog/helm-charts](https://github.com/DataDog/helm-charts), with raw manifests
+stored in the application repository generated with a [script](https://github.com/DataDog/datadog-agent/blob/main/Dockerfiles/manifests/generate.sh).
+
+## Drawbacks
+
+Extracting our Helm charts into their own repository adds additional management overhead, as well as possible context
+switching for contributors. Specifically it creates bidirectional dependencies between the application and chart repositories
+as we need to keep the `appVersion` in sync with Vector releases, and the Kubernetes manifests in sync with the chart releases.
+
+Triggering cross-repository workflows in GitHub Actions isn't well supported but appears possible. This could impact our
+ability to automate updates between the application and chart repositories.
+
+## Alternatives
+
+### Do Nothing
+
+Our existing setup forces our testing strategy to rely on costly compilation and building steps for even minor changes,
+doing nothing continues this inefficiency as we expand our testing requirements to cover a wider spread of uses.
+
+### Version charts inside of `timberio/vector`
+
+We could create unique tags for chart versions, `chart-agent-1.0.0` for example, and keep all charts in the application
+repository. This feels _hacky_ but likely would resolve the pain points with minimal adjustments to our existing workflows.
+
+## Outstanding Questions
+
+- Do we need to automate the cross-repository management or just document the steps with the initial work?
+- Do we need to migrate old chart releases to the `DataDog/helm-charts` repository, or direct users to the old repository for old releases?
```

Comment:
A few additional questions:

1. The charts are currently licenced under MPL 2.0, the DataDog repository uses Apache Licence 2.0. Is there anything we need to do here?
2. Do we need to update existing contributing docs
3. Is there any CLA that contributors need to sign when we migrate to the DD repo (I don't believe so, couldn't find anything)?
4. Will we keep the same PR/merge access when making chart related changes, or do we require approval from non-Vector members to merge PRs?

Reply:
> 1. The charts are currently licenced under MPL 2.0, the DataDog repository uses Apache Licence 2.0. Is there anything we need to do here?

Ah, good call - I didn't actually check, @binarylogic do you have any thoughts here. I'll check with the container-integrations team in the meantime.

> 2. Do we need to update existing contributing docs

I believed we would assume Datadog's [existing contributing docs](https://github.com/DataDog/helm-charts/blob/master/CONTRIBUTING.md) but could call that out specifically.

> 3. Is there any CLA that contributors need to sign when we migrate to the DD repo (I don't believe so, couldn't find anything)?

No CLA or DCO sort of thing needed.

> 4. Will we keep the same PR/merge access when making chart related changes, or do we require approval from non-Vector members to merge PRs?

We'll get access to the repository and get codeowner status on our charts 👍

Reply:
Just following up that we decided on keeping a separate repo in the `timberio` org for now which I think obviates some of these questions:

> * The charts are currently licenced under MPL 2.0, the DataDog repository uses Apache Licence 2.0. Is there anything we need to do here?

We can keep our MPL 2.0 license

> * Do we need to update existing contributing docs

It does seem like we'll need to do this to point users to in in our `CONTRIBUTING.md`

> * Is there any CLA that contributors need to sign when we migrate to the DD repo (I don't believe so, couldn't find anything)?

We can avoid this for now.

> * Will we keep the same PR/merge access when making chart related changes, or do we require approval from non-Vector members to merge PRs?

Same, we'll use our current processes.

> - Do we need to automate the cross-repository management or just document the steps with the initial work?

Could you expand a bit on what this means? I feel like I'm lacking some context.

> - Do we need to migrate old chart releases to the `DataDog/helm-charts` repository, or direct users to the old repository for old releases?

Is this still an issue if we keep it in the `timberio` org for now? Would it be the same helm chart repository?

Reply:
> > * Do we need to automate the cross-repository management or just document the steps with the initial work?
> 
> Could you expand a bit on what this means? I feel like I'm lacking some context.

Yep! When we cut a new Vector release. we'll want to update the Helm chart to use the image and in the opposite direction, if we make material changes to the charts that we need to regenerate the `distribution/kubernetes` contents. Does that make sense? It's probably possible to automate these, but otherwise can be scripts/make targets that we run when needed.

> > * Do we need to migrate old chart releases to the `DataDog/helm-charts` repository, or direct users to the old repository for old releases?
> 
> Is this still an issue if we keep it in the `timberio` org for now? Would it be the same helm chart repository?

That's a good question - we _could_ update our custom release script and continue using the existing s3 repo. If we wanted to adopt [chart-releaser](https://github.com/helm/chart-releaser) we'd end up with a new repo url and would need to consider migrating.

Reply:
> Yep! When we cut a new Vector release. we'll want to update the Helm chart to use the image and in the opposite direction, if we make material changes to the charts that we need to regenerate the `distribution/kubernetes` contents. Does that make sense? It's probably possible to automate these, but otherwise can be scripts/make targets that we run when needed.
> 
> > >

Ah, I see. I think we can start with just documenting the procedure and then automating it if it becomes a pain.



> That's a good question - we _could_ update our custom release script and continue using the existing s3 repo. If we wanted to adopt [chart-releaser](https://github.com/helm/chart-releaser) we'd end up with a new repo url and would need to consider migrating.

As long as we don't have to have everyone migrate again if/when we rename the `timberio` org then I think that is ok to ask them to migrate once as part of this cut over. Regarding old chart releases, would it be difficult to copy them?

Reply:
> As long as we don't have to have everyone migrate again if/when we rename the timberio org then I think that is ok to ask them to migrate once as part of this cut over. Regarding old chart releases, would it be difficult to copy them?

If we re-released old charts we could handle user migrations with redirects... We could do something _weird_ by doing tags/releases and have the git history still... or manually publish them but not have them in the git history